### ISSUE TITLE
L430 Compatibility

### ISSUE THREAD

**Issue by FouzR** [n4ru/1vyrain#87]
I tried the BIOS USB, it showed that it's compatible
With the below info
Machine:L430
BIOS:G3ETB6WW(2.76)
Version:
Flashsize:

Would I be able to use the whitelist scripts from here: https://github.com/n4ru/1vyrain/issues/27

**Comment by digmorepaka** [comment]
Currently we  don't have patches for L430/L530. Testing is needed.

**Comment by FouzR** [comment]
I see
I do have one question, my L430 has 2 miniPCIe slots, so would the whitelist be active on the secondary slot?

**Comment by unrelentingtech** [comment]
I currently have an L430 and a chip clip, I can test things.

So far I've tried `me_cleaner -S`, now it doesn't fully boot (hangs on `_` when entering setup or continuing boot after the RTC error) :D maybe this is that ME delay everyone's talking about and I wasn't patient enough.

Also notably flashrom doesn't seem to handle the IFD well on this? At least dumping `--ifd -i bios` or `--ifd -i me` or whatever also results in an 8MiB file (!) despite the expected "Using region" message.

<details>
<summary>ifdtool -d of the dumped stock rom</summary>

```
File l430.stock.rom is 8388608 bytes
ICH Revision: 6 series Cougar Point
FLMAP0:    0x02040003
  NR:      2
  FRBA:    0x40
  NC:      1
  FCBA:    0x30
FLMAP1:    0x12100206
  ISL:     0x12
  FPSBA:   0x100
  NM:      2
  FMBA:    0x60
FLMAP2:    0x00210120
  PSL:     0x2101
  FMSBA:   0x200
FLUMAP1:   0x00000cdf
  Intel ME VSCC Table Length (VTL):        12
  Intel ME VSCC Table Base Address (VTBA): 0x000df0

ME VSCC table:
  JID0:  0x001740ef
    SPI Component Vendor ID:            0xef
    SPI Component Device ID 0:          0x40
    SPI Component Device ID 1:          0x17
  VSCC0: 0x20052005
    Lower Erase Opcode:                 0x20
    Lower Write Enable on Write Status: 0x50
    Lower Write Status Required:        No
    Lower Write Granularity:            64 bytes
    Lower Block / Sector Erase Size:    4KB
    Upper Erase Opcode:                 0x20
    Upper Write Enable on Write Status: 0x50
    Upper Write Status Required:        No
    Upper Write Granularity:            64 bytes
    Upper Block / Sector Erase Size:    4KB
  JID1:  0x0017ba20
    SPI Component Vendor ID:            0x20
    SPI Component Device ID 0:          0xba
    SPI Component Device ID 1:          0x17
  VSCC1: 0x20052005
    Lower Erase Opcode:                 0x20
    Lower Write Enable on Write Status: 0x50
    Lower Write Status Required:        No
    Lower Write Granularity:            64 bytes
    Lower Block / Sector Erase Size:    4KB
    Upper Erase Opcode:                 0x20
    Upper Write Enable on Write Status: 0x50
    Upper Write Status Required:        No
    Upper Write Granularity:            64 bytes
    Upper Block / Sector Erase Size:    4KB
  JID2:  0x001720c2
    SPI Component Vendor ID:            0xc2
    SPI Component Device ID 0:          0x20
    SPI Component Device ID 1:          0x17
  VSCC2: 0x20052005
    Lower Erase Opcode:                 0x20
    Lower Write Enable on Write Status: 0x50
    Lower Write Status Required:        No
    Lower Write Granularity:            64 bytes
    Lower Block / Sector Erase Size:    4KB
    Upper Erase Opcode:                 0x20
    Upper Write Enable on Write Status: 0x50
    Upper Write Status Required:        No
    Upper Write Granularity:            64 bytes
    Upper Block / Sector Erase Size:    4KB
  JID3:  0x0017701c
    SPI Component Vendor ID:            0x1c
    SPI Component Device ID 0:          0x70
    SPI Component Device ID 1:          0x17
  VSCC3: 0x20052005
    Lower Erase Opcode:                 0x20
    Lower Write Enable on Write Status: 0x50
    Lower Write Status Required:        No
    Lower Write Granularity:            64 bytes
    Lower Block / Sector Erase Size:    4KB
    Upper Erase Opcode:                 0x20
    Upper Write Enable on Write Status: 0x50
    Upper Write Status Required:        No
    Upper Write Granularity:            64 bytes
    Upper Block / Sector Erase Size:    4KB
  JID4:  0x001740ef
    SPI Component Vendor ID:            0xef
    SPI Component Device ID 0:          0x40
    SPI Component Device ID 1:          0x17
  VSCC4: 0x20052005
    Lower Erase Opcode:                 0x20
    Lower Write Enable on Write Status: 0x50
    Lower Write Status Required:        No
    Lower Write Granularity:            64 bytes
    Lower Block / Sector Erase Size:    4KB
    Upper Erase Opcode:                 0x20
    Upper Write Enable on Write Status: 0x50
    Upper Write Status Required:        No
    Upper Write Granularity:            64 bytes
    Upper Block / Sector Erase Size:    4KB
  JID5:  0x0018ba20
    SPI Component Vendor ID:            0x20
    SPI Component Device ID 0:          0xba
    SPI Component Device ID 1:          0x18
  VSCC5: 0x20052005
    Lower Erase Opcode:                 0x20
    Lower Write Enable on Write Status: 0x50
    Lower Write Status Required:        No
    Lower Write Granularity:            64 bytes
    Lower Block / Sector Erase Size:    4KB
    Upper Erase Opcode:                 0x20
    Upper Write Enable on Write Status: 0x50
    Upper Write Status Required:        No
    Upper Write Granularity:            64 bytes
    Upper Block / Sector Erase Size:    4KB

OEM Section:
00: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
10: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
20: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
30: ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff

Found Region Section
FLREG0:    0x00000000
  Flash Region 0 (Flash Descriptor): 00000000 - 00000fff
FLREG1:    0x07ff0400
  Flash Region 1 (BIOS): 00400000 - 007fffff
FLREG2:    0x03ff0001
  Flash Region 2 (Intel ME): 00001000 - 003fffff
FLREG3:    0x00001fff
  Flash Region 3 (GbE): 00fff000 - 00000fff (unused)
FLREG4:    0x00001fff
  Flash Region 4 (Platform Data): 00fff000 - 00000fff (unused)

Found Component Section
FLCOMP     0x09300024
  Dual Output Fast Read Support:       not supported
  Read ID/Read Status Clock Frequency: 33MHz
  Write/Erase Clock Frequency:         33MHz
  Fast Read Clock Frequency:           33MHz
  Fast Read Support:                   supported
  Read Clock Frequency:                20MHz
  Component 2 Density:                 8MB
  Component 1 Density:                 8MB
FLILL      0x00000000
  Invalid Instruction 3: 0x00
  Invalid Instruction 2: 0x00
  Invalid Instruction 1: 0x00
  Invalid Instruction 0: 0x00
FLPB       0x00000000
  Flash Partition Boundary Address: 0x000000

Found PCH Strap Section
PCHSTRP0  : 0x0820d682
PCHSTRP1  : 0x0000010f
PCHSTRP2  : 0x00560000
PCHSTRP3  : 0x00000000
PCHSTRP4  : 0x00c8e000
PCHSTRP5  : 0x00000000
PCHSTRP6  : 0x00000000
PCHSTRP7  : 0x00000000
PCHSTRP8  : 0x00000000
PCHSTRP9  : 0x30000580
PCHSTRP10 : 0x00410044
PCHSTRP11 : 0x95000097
PCHSTRP12 : 0x00000000
PCHSTRP13 : 0x00000000
PCHSTRP14 : 0x00000000
PCHSTRP15 : 0x0000033e
PCHSTRP16 : 0x00000000
PCHSTRP17 : 0x00000002
PCHSTRP18 : 0xffffffff
PCHSTRP19 : 0xffffffff
PCHSTRP20 : 0xffffffff
PCHSTRP21 : 0xffffffff
PCHSTRP22 : 0xffffffff
PCHSTRP23 : 0xffffffff
PCHSTRP24 : 0xffffffff
PCHSTRP25 : 0xffffffff
PCHSTRP26 : 0xffffffff
PCHSTRP27 : 0xffffffff
PCHSTRP28 : 0xffffffff
PCHSTRP29 : 0xffffffff
PCHSTRP30 : 0xffffffff
PCHSTRP31 : 0xffffffff
PCHSTRP32 : 0xffffffff
PCHSTRP33 : 0xffffffff
PCHSTRP34 : 0xffffffff
PCHSTRP35 : 0xffffffff
PCHSTRP36 : 0xffffffff
PCHSTRP37 : 0xffffffff
PCHSTRP38 : 0xffffffff
PCHSTRP39 : 0xffffffff
PCHSTRP40 : 0xffffffff
PCHSTRP41 : 0xffffffff
PCHSTRP42 : 0xffffffff
PCHSTRP43 : 0xffffffff
PCHSTRP44 : 0xffffffff
PCHSTRP45 : 0xffffffff
PCHSTRP46 : 0xffffffff
PCHSTRP47 : 0xffffffff
PCHSTRP48 : 0xffffffff
PCHSTRP49 : 0xffffffff
PCHSTRP50 : 0xffffffff
PCHSTRP51 : 0xffffffff
PCHSTRP52 : 0xffffffff
PCHSTRP53 : 0xffffffff
PCHSTRP54 : 0xffffffff
PCHSTRP55 : 0xffffffff
PCHSTRP56 : 0xffffffff
PCHSTRP57 : 0xffffffff
PCHSTRP58 : 0xffffffff
PCHSTRP59 : 0xffffffff
PCHSTRP60 : 0xffffffff
PCHSTRP61 : 0xffffffff
PCHSTRP62 : 0xffffffff
PCHSTRP63 : 0xffffffff
PCHSTRP64 : 0x00000000
PCHSTRP65 : 0xffffffff
PCHSTRP66 : 0xffffffff
PCHSTRP67 : 0xffffffff
PCHSTRP68 : 0xffffffff
PCHSTRP69 : 0xffffffff
PCHSTRP70 : 0xffffffff
PCHSTRP71 : 0xffffffff
AltMeDisable bit is not set

Found Master Section
FLMSTR1:   0x0a0b0000 (Host CPU/BIOS)
  Platform Data Region Write Access: disabled
  GbE Region Write Access:           enabled
  Intel ME Region Write Access:      disabled
  Host CPU/BIOS Region Write Access: enabled
  Flash Descriptor Write Access:     disabled
  Platform Data Region Read Access:  disabled
  GbE Region Read Access:            enabled
  Intel ME Region Read Access:       disabled
  Host CPU/BIOS Region Read Access:  enabled
  Flash Descriptor Read Access:      enabled
  Requester ID:                      0x0000

FLMSTR2:   0x0c0d0000 (Intel ME)
  Platform Data Region Write Access: disabled
  GbE Region Write Access:           enabled
  Intel ME Region Write Access:      enabled
  Host CPU/BIOS Region Write Access: disabled
  Flash Descriptor Write Access:     disabled
  Platform Data Region Read Access:  disabled
  GbE Region Read Access:            enabled
  Intel ME Region Read Access:       enabled
  Host CPU/BIOS Region Read Access:  disabled
  Flash Descriptor Read Access:      enabled
  Requester ID:                      0x0000

FLMSTR3:   0x08080118 (GbE)
  Platform Data Region Write Access: disabled
  GbE Region Write Access:           enabled
  Intel ME Region Write Access:      disabled
  Host CPU/BIOS Region Write Access: disabled
  Flash Descriptor Write Access:     disabled
  Platform Data Region Read Access:  disabled
  GbE Region Read Access:            enabled
  Intel ME Region Read Access:       disabled
  Host CPU/BIOS Region Read Access:  disabled
  Flash Descriptor Read Access:      disabled
  Requester ID:                      0x0118

Found Processor Strap Section
????:      0x00000000
????:      0xffffffff
????:      0xffffffff
????:      0xffffffff
????:      0xffffffff
????:      0xffffffff
????:      0xffffffff
????:      0xffffffff
```

</details>

What should I test first? Putting https://github.com/n4ru/patched-bioses/blob/master/L430.rom into the `bios` partition and flashing externally?

**Comment by unrelentingtech** [comment]
So, more testing!

Injecting [that rom](https://github.com/n4ru/patched-bioses/blob/b91aaf3febc871a23201933c718df0777be267c1/L430.rom) with `ifdtool` results in not booting at all.

Applied patches to the stock `bios` region as extracted with `ifdtool` (firmware `G3ET65WW(2.04)` from 2012-09-25):

- Advanced menu works (presumably overclock too but I haven't tested that)
- [These patches](https://github.com/n4ru/1vyrain/issues/27#issuecomment-595051757) did not actually bypass the Wi-Fi check :(

**Comment by unrelentingtech** [comment]
So one condition was missed in those patches, the full working patch set is:

```
79E0EDD7-9D1D-4F41-AE1A-F896169E5216 10 P:C8390E0F84:C8390E90E9 
79E0EDD7-9D1D-4F41-AE1A-F896169E5216 10 P:C8390E75620F:C8390E75000F 
79E0EDD7-9D1D-4F41-AE1A-F896169E5216 10 P:C8394E04754E83:C8394E04750083
79E0EDD7-9D1D-4F41-AE1A-F896169E5216 10 P:0983FA0474:0983FA04EB
```

Now that boots with a ralink Wi-Fi card \o/

**Comment by digmorepaka** [comment]
Could you submit a PR in https://github.com/digmorepaka/thinkpad-firmware-patches/ with all the details? Afterwards it will be added to 1vyrain and will need one final test before release.

**Comment by unrelentingtech** [comment]
Yeah, just found that repo while reading the other threads :)

> will need one final test before release

Especially considering how the .rom from the patched-bioses repo didn't boot at all for me. I'll have to revert to stock and try flashing from the live USB, right? To test the full "end user" experience.

**Comment by digmorepaka** [comment]
@unrelentingtech Alright this is a binary of the latest UEFI with the patchset. You will need to pad it by 4M and flash it with the --ifd -i bios parameter manually with flashrom(https://github.com/n4ru/1vyrain/issues/64#issuecomment-761773042)


http://0x0.st/-7IC.fixed
`e900453fa85235eab78c40e629b35134`

**Comment by unrelentingtech** [comment]
It works \o/

(I've also tried to test the TPM… it's in MFG mode, clearing from Windows does nothing, and I couldn't get tpm-tools on Arch because github is having issues lol)

**Comment by digmorepaka** [comment]
MFG mode means completely off, try this, it should either do 5 beeps and still not work or work perfectly fine

http://0x0.st/-7U-.fixed

`1d3cead051fd59f5fb377c25f76e3472` md5

**Comment by unrelentingtech** [comment]
Works great, what was changed, just self-signing?

**Comment by digmorepaka** [comment]
For some reason the MFG mode was turned on by thinkpad-uefi-sign(it shouldn't do that). 

This is a full image, so this should autodetect your model. https://xorg-broke-aga.in/1vyrain.iso md5 `50b763065b73443e68e34eff24c36695`

**Comment by unrelentingtech** [comment]
Looks like you've added the rom file but didn't uncomment the start.sh lines.

**Comment by digmorepaka** [comment]
It appears that i have uploaded the wrong file. Fixed, same link md5sum `43c896ecf7b2b368020be7ac263af301`

**Comment by unrelentingtech** [comment]
The same link still returns `50b763065b73443e68e34eff24c36695`…

**Comment by digmorepaka** [comment]
ergh cloudflare cache probably not updated yet,I did check  the file on the server and it is correct. I issued a cache purge so it *should* be the right file

**Comment by unrelentingtech** [comment]
`line 49: [: : integer expression expected`

hmm seems like the version extraction command fails. Running it manually, the awk invocation extracts nothing from `$3` — really `$2` is `G3ET65WW(2.04)`. Do newer firmwares / other models always have whitespace before the opening `(`?

**Comment by digmorepaka** [comment]
Looking over some other reports, it does look like other models use a different format and we have trouble detecting L430. I do have an idea for a bandaid fix as the latest(which is extremely likely final) UEFI still have the vulnerability. I will work on  it tomorrow

**Comment by unrelentingtech** [comment]
Well, one possibility is using a regexp for the parentheses on the whole line without using whitespace "fields"…

**Comment by FouzR** [comment]
So, by the looks of things, you'll update the iso by tomorrow with L430 support?

**Comment by digmorepaka** [comment]
@FouzR you can already flash the rom manually, but for full automatic support that is the plan. https://github.com/n4ru/1vyrain/issues/87#issuecomment-980782028

**Comment by FouzR** [comment]
I'll wait for the full automatic support just to be safe :)

**Comment by digmorepaka** [comment]
@unrelentingtech same link, `588896ad851ec9d5fffb914e205840c6` bypassed version check as it's not needed

**Comment by unrelentingtech** [comment]
Now everything works perfectly! \o/

**Comment by FouzR** [comment]
On which BIOS version did you try it?

**Comment by FouzR** [comment]
![4F11B425-AB2D-4DDD-98EB-233829437BFE](https://user-images.githubusercontent.com/88454554/144017376-2a1e4216-018c-4625-a626-f20fced2e9b1.jpeg)
It's showing this when I tried to do on mine

**Comment by digmorepaka** [comment]
PITA I know what this is, lucky that we caught it before releasing.

**Comment by FouzR** [comment]
Was that due to the band-aid fix you implemented?

**Comment by digmorepaka** [comment]
Nah, this stems from Macronix chips not being autodetectable by software like flashrom. Good thing our model only ships with one of those(and in the case that someone replaced it there are failsafes). It will  be fixed like it was fixed on T430s, not at my PC at the moment so you gotta  wait.

**Comment by FouzR** [comment]
Cool, I'm happy you guys are doing this, so, take your time I guess :)

**Comment by unrelentingtech** [comment]
Interesting, my unit came with a Winbond chip

**Comment by digmorepaka** [comment]
@FouzR md5: `4be31f563abd41cc74749ba03b21e713` https://xorg-broke-aga.in/1vyrain.iso it should work now.

Just for  future reference i also added L530  support, UEFI is shared between L530 and L430 so I hope the autodetection works

**Comment by FouzR** [comment]
![IMG_0632](https://user-images.githubusercontent.com/88454554/144154567-33ffd20b-8433-43d6-817b-8f5adc0db3d1.JPG)
Uhhhh

**Comment by digmorepaka** [comment]
@unrelentingtech you were able to reflash from the latest version yeah? Could you check the latest iso I posted on your board with a winbond and latest UEFI?

**Comment by unrelentingtech** [comment]
No, I reverted to stock every time, flashrom failed to write (but seemingly didn't corrupt anything) when already modded. (That was tested before auto detection worked)

**Comment by FouzR** [comment]
@unrelentingtech did you try the iso at the latest BIOS(2.76)?

**Comment by unrelentingtech** [comment]
No, not at stock latest bios, will test soon

**Comment by leecher1337** [comment]
Sorry for hijacking this thread, but it seems that there are persons who have
the capability to reflash a damaged BIOS ROM in the event that something goes
wrong.
As I do not have this facility, but successfully created and tested a [patch](https://github.com/hamishcoleman/thinkpad-ec/issues/203#issuecomment-991820190)
for a 7-row keyboard for the L430, I'm a bit hesitant to flash the BIOS, as I 
do not want to brick the machine.
Therefore I would like to ask if it is possible that you can test if
updating the EC Firmware would cause problems possibly with some checksums,
or if it is enough to just apply the patched EC firmware to the BIOS image and
reflash via flashrom.
This would be very helpful. In order to carry out the test, I'd request you to
do a harmless patch of the EC keyboard map, which doesn't have an effect on
the proper operation of the keyboard, but shows if changing EC firmware would
cause some CRC error to go off.

The EC firmware is divided into 2 parts. The first part is the loader that 
loads attached real firmware image to address 20100.
As only the loader-Part of the firmware is checksummed, it is not necessary
needed to recalculate the checksum of the loader part.

So for testing, the following steps should be carried out.

1) Flash most recent stock firmware with EC Firmware version Firmware G3HT40WW(1.14)
2) Dump BIOS with:

   `flashrom -p internal -r current-bios.bin --ifd -i bios`

3) Check BIOS image so verify that we are patching the correct region

   `dd if=current-bios.bin bs=1 count=16 skip=$[0x41E5BD] 2>/dev/null | hexdump -C`

   This should result in the first row of the keyboard map:

   `00000000  0e 05 0a 2e 36 55 16 4e  18 00 a2 a1 00 00 00 8c  |....6U.N........|`

4) Patch an empty key mapping at offset C for test purposes:

   ```
   cp current-bios.bin new-bios.bin
   echo -ne "\x18" | dd conv=notrunc of=new-bios.bin bs=1 seek=$[0x41E5C9]
   ```

5) Flash back BIOS to machine

   `flashrom -p internal -w new-bios.bin --ifd -i bios`

6) Shutdown machine, remove power connection and battery for at least 30 seconds so
   that embedded controller is forced to reinitialize

7) Power up machine, install and load [x2100-ec-sys kernel module](https://github.com/exander77/x2100-ec-sys)

   `modprobe x2100_ec_sys`

8) Check memory region if the byte at offset C has changed:

   `dd if=/sys/kernel/debug/ec/ec0/ram bs=1 count=16 skip=$[0x366BD] 2>/dev/null | hexdump -C`

    Output should now read:

   `00000000  0e 05 0a 2e 36 55 16 4e  18 00 a2 a1 18 00 00 8c  |....6U.N........|`


I would be very grateful if someone can check this so that I can permanently
install the 7-row keyboard firmware fix then.

Thank you very much in advance!

**Comment by digmorepaka** [comment]
@leecher1337  This is unrelated to this issue, please do not clutter up the thread. Message the people directly, and or bring them to the ec repo issue. https://github.com/hamishcoleman/thinkpad-ec/issues/203#issuecomment-991820190

**Comment by unrelentingtech** [comment]
So it just fails to flash on 2.76, unsurprisingly doesn't matter whether modded or not.

Looks like `if [ "255" -gt "$version" ];` was the correct check — flashing does work on 2.54. So don't delete the version check, just make it work whether there's whitespace before the `(` or not.

**Comment by unrelentingtech** [comment]
@digmorepaka so what was the fix for

> MFG mode was turned on by thinkpad-uefi-sign

? I'd like to sign my own builds in a way that makes the TPM work…

**Comment by digmorepaka** [comment]
> ? I'd like to sign my own builds in a way that makes the TPM work…

It seems that running thinkpad-uefi-sign and then manually changing 4C 4E 56 42 42 53 45 43 FF to 4C 4E 56 42 42 53 45 43 FB did the trick.

**Comment by snakeeater4526** [comment]
Hi there, i just read this thread and i have a L530 , does that mean that 1vyrain support L530 now?

**Comment by digmorepaka** [comment]
UEFI itself is identical, not yet supported in 1vyrain due to different model string, but UEFI patch is working.

**Comment by devbrones** [comment]
Hey! I have a L430 and soon a chip clip, can maybe do some testing if time allows it.
Is 1vyrain stable on L430 and if so how can i install it?

**Comment by devbrones** [comment]
> Hey! I have a L430 and soon a chip clip, can maybe do some testing if time allows it. Is 1vyrain stable on L430 and if so how can i install it?
> 
> my bios version is G3ET65WW(2.04) and i have a Winbond W25Q64.V rom

update btw, worked flawlessly on the standard iso from 1vyra.in

**Comment by BoredOperator** [comment]
Hello i tried to run the 1vyrain.iso on my Thinkpad L430 (with the bios versions 2.76, 2.54 and 2.04) and each time i get the message that my machine is compatible but unsupported. (I dont have a hardware flasher nor do i have any flashing experience) Is there a way to enable support?

Machine: L430
BIOS: G3ET94WW(2.54)
Version:
Flashsize: 8192

**Comment by digmorepaka** [comment]
> Hello i tried to run the 1vyrain.iso on my Thinkpad L430 (with the bios versions 2.76, 2.54 and 2.04) and each time i get the message that my machine is compatible but unsupported. (I dont have a hardware flasher nor do i have any flashing experience) Is there a way to enable support?
> 
> Machine: L430 BIOS: G3ET94WW(2.54) Version: Flashsize: 8192

Yeah it's currently in a state where it is supported but not really. You can flash internally, rom is made, but the version string decoding hasn't been done yet because it requires quite a rework to support a total of two laptops that a minority of people have.

To flash your machine(internet connection required)
1. go through the compatibility check, after it wakes up again and says flashing is possible, exit the script with Ctrl+C
2. `wget https://github.com/n4ru/patched-bioses/raw/refs/heads/master/L430.rom`
3. `dd if=/dev/zero of=pad.bin bs=1M count=4` 
4. `cat pad.bin L430.rom > UEFI.rom`
5. `/root/flashrom/flashrom -p internal:laptop=force_I_want_a_brick -w UEFI.rom --ifd -i bios -N`
after that is done, you can reboot

**Comment by BoredOperator** [comment]
Hello again,
my brother has recently bought the same laptop model for cheap and wanted me to flash his one as well. When i follow the same process i get the same error Message @FouzR encountered before. It seems it does not detect the Macronix chips. Which one do i need to flash with the -c option? 

His Laptop also does not report FLASHSIZE when running the iso 

Machine: L430 
BIOS: G3ET94WW(2.54) 
![PXL_20241210_152740009](https://github.com/user-attachments/assets/43851696-76d5-42da-b56d-b49c333bbe82)

Version: 
Flashsize: 
![PXL_20241210_152740009](https://github.com/user-attachments/assets/f386b74c-c490-43c0-9b17-6d2c58d96d2c)

**Comment by digmorepaka** [comment]
> It seems it does not detect the Macronix chips. Which one do i need to flash with the -c option?

Sorry for the late reply, probably wont be relevant for you anymore but if anyone is manually flashing it and encounters this, on the L430 it can only be a W25Q64(CVSSIG) - autodetect works on this, MX25L6404E(EM2I-12G) - flashrom cannot autodetect these, you will have to use the `-c "MX25L6406/MX25L6408E"` parameter for flashrom, and the last IC that was used was Numynox N25W064A(13ESE40F), no idea if autodetection works on the numynox.

**Comment by ghost** [comment]
How is the support for this laptop? And what versions of the BIOS I need? (Lenovo Thinkpad L430, right now on 2.76)

**Comment by ghost** [comment]
Hi, so I installed this and it worked perfectly! Thanks for the instructions although it sucks that it's manual.

**Comment by snakeeater4526** [comment]
comfirmed to work on L530 after the custom ec have been installed
